docs/examples/Custom_Pandas_Dataloader.py

```diff
@@ -0,0 +1,280 @@
+    time_field_format: str = Property(
+        default=None, doc='Optional datetime format')
+    timestamp: bool = Property(
+        default=False, doc='Treat time field as a timestamp from epoch')
+    metadata_fields: Collection[str] = Property(
+        default=None, doc='List of columns to be saved as metadata, default all')
+
+    def _get_metadata(self, row):
+        if self.metadata_fields is None:
+            local_metadata = dict(row)
+            for key in list(local_metadata):
+                if key == self.time_field or key in self.state_vector_fields:
+                    del local_metadata[key]
+        else:
+            local_metadata = {field: row[field]
+                              for field in self.metadata_fields
+                              if field in row}
+        return local_metadata
+
+    def _get_time(self, row):
+        if self.time_field_format is not None:
+            time_field_value = datetime.strptime(row[self.time_field], self.time_field_format)
+        elif self.timestamp is True:
+            fractional, timestamp = modf(float(row[self.time_field]))
+            time_field_value = datetime.utcfromtimestamp(int(timestamp))
+            time_field_value += timedelta(microseconds=fractional * 1E6)
+        else:
+            time_field_value = parse(row[self.time_field], ignoretz=True)
+        return time_field_value
+
+
+# %%
+# Data Ground Truth Reader
+# ^^^^^^^^^^^^^^^^^^^^^^^^
+# With the help of our `_DataFrameReader` class, we can now define a custom
+# `DataFrameGroundTruthReader`. This is similar to :class:`~.CSVGroundTruthReader` and 
+# inherits from the base `GroundTruthReader` class. A key difference is that we 
+# include an instance attribute for the dataframe containing our data.
+# 
+# We also define a custom generator function (groundtruth_paths_gen) that uses the decorator
+# `@BufferedGenerator.generator_method`. The generator needs to return a time and a set of
+# detections, like so:
+
+class DataFrameGroundTruthReader(GroundTruthReader, _DataFrameReader):
+    """A custom reader for pandas DataFrames containing truth data.
+
+    The DataFrame must have colums containing all fields needed to generate the
+    ground truth state. Those states with the same ID will be put into
+    a :class:`~.GroundTruthPath` in sequence, and all paths that are updated at the same time
+    are yielded together, and such assumes file is in time order.
+
+    Parameters
+    ----------
+    """
+    dataframe: pd.DataFrame = Property(doc="DataFrame containing the ground truth data.")
+    path_id_field: str = Property(doc='Name of column to be used as path ID')
+
+    @BufferedGenerator.generator_method
+    def groundtruth_paths_gen(self):
+        """ Generator method for providing each row of ground truth data. """
+        groundtruth_dict = {}
+        updated_paths = set()
+        previous_time = None
+        for row in self.dataframe.to_dict(orient="records"):
```

Comment:
I'm sure we could do this easily enough, although I've not included it in the current commit yet though. Before doing this, I could do with confirming the precise details in order to avoid accidentally changing the current logic of the generator.

If we simply order by path_id and time, and then iteratively yield each time and updated_paths (detections for DataFrameDetectionReader), is that doing precisely the same functionality as the current groundtruth_paths_gen (and detections_gen)?